### PR TITLE
Fixes #2017 Move Initial Conditions relating to an expression profile…

### LINKED ISSUE #2017: Move Initial Conditions relating to an expression profile into the expression profile object
Related to https://github.com/Open-Systems-Pharmacology/MoBi/issues/855

### PR DESCRIPTION
**Pull request by rwmcintosh**
… into the expression profile object

### CODE REVIEW

File: src/OSPSuite.Core/Domain/Builder/SimulationBuilder.cs [hunk 1]
```diff
@@ -92,7 +92,8 @@ private void performMerge()
          _observers.AddRange(allBuilder(x => x.Observers));
          _molecules.AddRange(allBuilder(x => x.Molecules));
          _parameterValues.AddRange(allStartValueBuilder(x => x.SelectedParameterValues));
-         _initialConditions.AddRange(allStartValueBuilder(x => x.SelectedInitialConditions));
+         _initialConditions.AddRange(allStartValueBuilder(x => x.SelectedInitialConditions)
```
Comment (rwmcintosh) [inline src/OSPSuite.Core/Domain/Builder/SimulationBuilder.cs:95]: Now we need to look in all the expression profiles to build the list of all InitialConditions

File: src/OSPSuite.Core/Domain/Builder/ExpressionProfileBuildingBlock.cs [hunk 2]
```diff
@@ -37,6 +40,29 @@ public override string Name
...
+         initialCondition.BuildingBlock = this;
+      }
+
+      public override void AcceptVisitor(IVisitor visitor)
+      {
+         base.AcceptVisitor(visitor);
+         _initialConditions.Each(ic => ic.AcceptVisitor(visitor));
+      }
+
+      public IEnumerable<InitialCondition> InitialConditions => _initialConditions;
```
Comment (msevestre) [inline src/OSPSuite.Core/Domain/Builder/ExpressionProfileBuildingBlock.cs:64]: IReadonly?

File: src/OSPSuite.Core/Domain/Builder/ExpressionProfileBuildingBlock.cs [hunk 3]
```diff
@@ -48,6 +74,8 @@ public override void UpdatePropertiesFrom(IUpdatable source, ICloneManager clone
          // Name is required because our base objects will the private property
          // But in this case the name is decomposed and stored in 3 other properties
          Name = sourceExpressionProfile.Name;
+
+         sourceExpressionProfile.InitialConditions.Each(initialCondition => AddInitialCondition(cloneManager.Clone(initialCondition)));
```
Comment (msevestre) [inline src/OSPSuite.Core/Domain/Builder/ExpressionProfileBuildingBlock.cs:78]: clear your initial condition probably before doing this?

File: src/OSPSuite.Core/Domain/Builder/PathAndValueEntityBuildingBlock.cs [hunk 4]
```diff
@@ -50,7 +50,7 @@ IEnumerator IEnumerable.GetEnumerator()
       public override void AcceptVisitor(IVisitor visitor)
       {
          base.AcceptVisitor(visitor);
-         _allValues.Each(initialCondition => initialCondition.AcceptVisitor(visitor));
+         _allValues.Each(pathAndValueEntity => pathAndValueEntity.AcceptVisitor(visitor));
```
Comment (msevestre) [inline src/OSPSuite.Core/Domain/Builder/PathAndValueEntityBuildingBlock.cs:53]: Personal thoughts for discussion: not sure big long name in lambdas make it easier to read. Above you used ic, ok. x, for the win in this case (or v for value)

File: src/OSPSuite.Core/Domain/Builder/SimulationBuilder.cs [hunk 5]
```diff
@@ -92,7 +92,8 @@ private void performMerge()
...
+            .Concat(_simulationConfiguration.ExpressionProfiles.SelectMany(expressionProfile => expressionProfile.InitialConditions)));
```
Comment (msevestre) [inline src/OSPSuite.Core/Domain/Builder/SimulationBuilder.cs:96]: consistency. Why not use x here where we have x everywhere. It feels like it;s something difference and it's not

**Review by msevestre — CHANGES_REQUESTED** [review]

File: src/OSPSuite.Core/Domain/Builder/ExpressionProfileBuildingBlock.cs (hunk 2 above)
Comment (rwmcintosh) [reply src/OSPSuite.Core/Domain/Builder/ExpressionProfileBuildingBlock.cs:64]: It's converting a Cache, so I don't know if it's better or not. I would need to call ToList on it for each access. This is templated from the PathAndValueEntityBuildingBlock which does the same.

Comment (msevestre) [reply src/OSPSuite.Core/Domain/Builder/ExpressionProfileBuildingBlock.cs:64]: There is another thin. IReadonlyDictionary or sthg?

Comment (rwmcintosh) [reply src/OSPSuite.Core/Domain/Builder/ExpressionProfileBuildingBlock.cs:64]: image

Looks like IReadOnlyCollection is the one.

**Review by msevestre — APPROVED** [review]

**Pull request merged**